### PR TITLE
Typing Annotations

### PR DESCRIPTION
**Pull request by hardbyte**
Python 2 and 3 compatible type annotations for the mypy static type checker.

I've been testing with:
```
pip install mypy
mypy clkhash --ignore-missing-imports --strict-optional --no-implicit-optional --disallow-untyped-calls
```

### CODE REVIEW

File: clkhash/key_derivation.py [hunk 1]
```diff
@@ -71,28 +77,31 @@ def __init__(self, master_secret, salt=None, info=None, hash_algo='SHA256'):
 
     @staticmethod
     def check_is_bytes(value):
+        # type: (Union[str, bytes]) -> bytes
```
Comment (wilko77) [inline clkhash/key_derivation.py]: If you go by the strict definition of this function, then it should be able to accept any type. More like `# type: (Any) -> bytes`

If you look at the context in which we use it in, then your definition is more precise.

However, it kind of seems weird to only allow str and bytes.

File: clkhash/key_derivation.py [hunk 2]
```diff
@@ -71,28 +77,31 @@ def __init__(self, master_secret, salt=None, info=None, hash_algo='SHA256'):
...
         if isinstance(value, bytes):
             return value
         else:
             raise TypeError('provided value is not of type "bytes"')
 
     @staticmethod
     def check_is_bytes_or_none(value):
+        # type: (Optional[bytes]) -> Optional[bytes]
```
Comment (wilko77) [inline clkhash/key_derivation.py]: same here

File: clkhash/randomnames.py [hunk 3]
```diff
@@ -24,12 +26,17 @@
...
+    if data_bytes is None:
+        raise ValueError("No data resource found with name {}".format(resource_name))
+    else:
+        data = data_bytes.decode('utf8')
+        reader = csv.reader(data.splitlines())
+        next(reader, None)  # skip the headers
+        return [row[0] for row in reader]
 
 
 def save_csv(data, schema, file):
```
Comment (wilko77) [inline clkhash/randomnames.py]: why does this function not have a type annotation?

File: clkhash/key_derivation.py (hunk 1 above)
Comment (hardbyte) [reply clkhash/key_derivation.py]: Ok will change to `Any`. 

My approach was to try avoid changing code unless it was essential to make the type checker happy. Personally I'd refactor this into `is_bytes` returning a `bool`.

File: clkhash/randomnames.py (hunk 3 above)
Comment (hardbyte) [reply clkhash/randomnames.py]: I missed it, and none of the typed functions must call it (otherwise the typechecker would have told me off)

Comment (hardbyte) [reply clkhash/randomnames.py]: Or maybe I left it as an excercise for the careful reviewer...

**Comment by hardbyte** [comment]
@sjhardy Are you able to review this while @wilko77 is away? I think I've addressed all of his comments.

**Pull request merged**